### ISSUE TITLE
Shared subscription sticky strategy stops working after a rollout restart of cluster

### ISSUE THREAD

**Issue by hholstil** [emqx/emqx#13535]
### What happened?

Shared subscription sticky strategy stops working after a rollout restart of cluster, if the subscribers are reconnecting during the rollout. Attached a test setup for reproducing the issue.

[emqx-sticky-test.tar.gz](https://github.com/user-attachments/files/16412826/emqx-sticky-test.tar.gz)

# Steps to reproduce:

1. Run the example deployment:
```
$ ./build-images.sh
$ docker compose up -d
```

2. Verify log output from the shared subscribers.
```
$ docker logs emqx-sticky-test-shared-subscriber-1
$ docker logs emqx-sticky-test-shared-subscriber-2
$ docker logs emqx-sticky-test-shared-subscriber-3
```

They should contain a single set of "I am now the owner of ..." messages
for the publishers (according to how the publishers were distributed), but
no "ownership of ... changed to ...".

3. Stop the publishers

Edit `docker-compose.yaml` and change the replica count
of `test-publisher` to 0 and run `docker compose up -d` again.
Leave the subscribers running.

4. Restart the EMQX brokers

```
$ ./rollout-restart.sh
```

5. Verify that the cluster status is ok

```
$ docker exec emqx1 emqx ctl cluster status
$ docker exec emqx2 emqx ctl cluster status
$ docker exec emqx3 emqx ctl cluster status
```

5. Restart the publishers

Edit `docker-compose.yaml` and change the replica count
of `test-publisher` back to the original value and run
`docker compose up -d` again.

6. Observe the shared subscriber logs

```
$ docker logs emqx-sticky-test-shared-subscriber-1
$ docker logs emqx-sticky-test-shared-subscriber-2
$ docker logs emqx-sticky-test-shared-subscriber-3
```

There's now "ownership of ... changed to ..." messages; one can keep checking the logs
to see that more keep appearing as new messages are being published.

This would indicate that the new publishers are not getting properly stickied to the
subscribers.

7. Restart the subscribers

```
$ docker restart emqx-sticky-test-shared-subscriber-1 emqx-sticky-test-shared-subscriber-2 emqx-sticky-test-shared-subscriber-3
```

8. Observe the shared subscriber logs

No more "ownership of ... changed to ..." messages are appearing, indicating that the sticky strategy is working again as expected.

9. Clean up the test deployment

```
$ docker compose down -v
```

### What did you expect to happen?

Sticky strategy should work even after the rollout restart with the attached example.

### How can we reproduce it (as minimally and precisely as possible)?

See the attached test setup that works with docker compose.

### Anything else we need to know?

I have originally observed the issue when running the cluster on kubernetes. Managed to create a simple test setup on docker compose that reproduces it.

### EMQX version

Official docker image emqx/emqx:5.7.1

**Comment by id** [comment]
@hholstil have you tried to start subscriber with `clean_session = False`?

**Comment by id** [comment]
Right. Could you try to add `Session-Expiry-Interval = 300` in the `properties` field of the `connect` call? By default I believe session expiry interval is 0, meaning the session is deleted as soon as client disconnects.

**Comment by zmstone** [comment]
EMQX implements the sticky strategy by pairing up the publisher client (actor) and the subscriber client process (actor).
The pair is kept as long as both processes are alive.
To have a stronger sticky pair, the implementation has to be changed to use a global, replicated storage (something outside of the procees/actor lifecycle) --- which is more of a feature request.

**Comment by zmstone** [comment]
Hi @hholstil 
Sorry, I misunderstood the issue in the beginning. Marked it as a bug.

**Comment by zmstone** [comment]
> Maybe this could be fixed by repopulating ?ALIVE_SHARED_SUBSCRIBERS from ?SHARED_SUBSCRIPTION during init. After doing that I can no longer reproduce the issue with the attached test deployment, but I'm not knowledgeable enough to say if this would work reliably in all cases.

The pid should have been inserted back after the client reconnected to the nodes which rebooted.
Will try to dig deeper why this happned.

**Comment by zmstone** [comment]
maybe consider remove ALIVE_SHARED_SUBSCRIBERS all together.
it seems not necessary since remote pids will eventually be deleted from the shared table
using another table to record mote alive pids does not really make it faster to find another alive member sooner because it relies on the underlying mnesia data replication (or node monitoring) anyway. AND it also seemingly causing data discrepancy.

**Comment by zmstone** [comment]
Hi @hholstil Thank you very much for the detailed report and scripts to reproduce.
The fix has been merged to `master` branch, and it should be included in 5.8.1 release (planed release is in mid October 2024).
